proxy/grpcproxy/watch.go

```diff
@@ -71,13 +97,30 @@ func (wp *watchProxy) Watch(stream pb.Watch_WatchServer) (err error) {
 
 		watchCh: make(chan *pb.WatchResponse, 1024),
 
-		ctx:    ctx,
-		cancel: cancel,
+		proxyCtx: wp.ctx,
+		ctx:      ctx,
+		cancel:   cancel,
 	}
 
-	go sws.recvLoop()
-	sws.sendLoop()
-	return wp.ctx.Err()
+	// recv/send => terminate server stream
+	donec := make(chan struct{}, 2)
+	go func() {
+		defer func() { donec <- struct{}{} }()
+		sws.recvLoop()
+	}()
+	go func() {
+		defer func() { donec <- struct{}{} }()
+		sws.sendLoop()
+	}()
+	<-donec
```

Comment:
both aren't guaranteed to exit until the underlying stream contexts are canceled